src/geom/CoordinateSequence.cpp

```diff
@@ -35,16 +36,274 @@ namespace geom { // geos::geom
 static Profiler* profiler = Profiler::instance();
 #endif
 
+// If GEOS_COORDSEQ_PADZ is defined:
+// - XY sequences will be stored as XYZ
+// - XYM sequences will be stored as XYZM
+// This prevents incorrect results when an XYZ Coordinate is read from
+// a sequence storing XY or XYM. When GEOS is changed to check
+// coordinate types throughout the libary, this can be undefined to
+// store coordinates efficiently.
+#define GEOS_COORDSEQ_PADZ
+
+CoordinateSequence::CoordinateSequence() :
+    CoordinateSequence(0, 0) {}
+
+CoordinateSequence::CoordinateSequence(std::size_t sz, bool hasz, bool hasm, bool init) :
+#ifdef GEOS_COORDSEQ_PADZ
+    m_vect(sz * (3u + hasm)),
+    m_stride(static_cast<std::uint8_t>(3u + hasm)),
+#else
+    m_vect(sz * (2u + hasm + hasz)),
+    m_stride(static_cast<std::uint8_t>(2u + hasm + hasz)),
+#endif
+    m_hasdim(true),
+    m_hasz(hasz),
+    m_hasm(hasm)
+{
+    if (init) {
+        initialize();
+    }
+}
+
+CoordinateSequence::CoordinateSequence(std::size_t sz, std::size_t dim) :
+    m_vect(sz * 3),
+    m_stride(3u),
+    m_hasdim(dim > 0),
+    m_hasz(dim == 3),
+    m_hasm(false)
```

Comment:
Running into another issue in Shapely. 

If one would do `GEOSCoordSeq_create(1, 4)`, which ends up above `CoordinateSequence(1, 4)`, that would create a 3D coordseq with both hasz and hasm set to False. 

I see that the documentation of both `CoordinateSequence(std::size_t size, std::size_t dim = 0)` and the `GEOSCoordSeq_create` C API says that this is only for creating XY or XYZ sequences (so `dim` can be 2 or 3). So that's a user error that I pass a value of 4. 
But just noting that before, this would raise an exception, while now you silently get wrong values if you subsequently set the ordinates with `GEOSCoordSeq_setOrdinate_r` / `CoordinateSequence::setOrdinate` (the ordinate index 2 and 3 (z and m) will overwrite each other)

This is totally fixable on the Shapely side (we just need to verify the dimension on our side instead of just passing it to GEOS). But I do wonder if the `GEOSCoordSeq_create` should be updated to allow passing dim=4?

Reply:
Yes, see https://github.com/libgeos/geos/pull/753